### PR TITLE
Simple deduction of type

### PR DESCRIPTION
**Pull request by brinxmat**
Firstly, apologies about the size of this PR.

- I refactored SPARQL out of the code; I had a working variant for type extraction, but it became very large and complex because of the interplay between different elements in different places
- Type extraction is currently very naïve, so this will need to be improved upon

### CODE REVIEW

File: metadata-transformer/src/main/java/no/unit/nva/metadata/filters/FilterShorterTitles.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package no.unit.nva.metadata.filters;
+
+import no.unit.nva.metadata.extractors.ExtractorUtil;
+import no.unit.nva.metadata.type.DcTerms;
+import org.eclipse.rdf4j.model.Literal;
+import org.eclipse.rdf4j.model.Model;
+import org.eclipse.rdf4j.model.Statement;
+import org.eclipse.rdf4j.model.Value;
+
+public final class FilterShorterTitles {
```
Comment (axthosarouris) [inline metadata-transformer/src/main/java/no/unit/nva/metadata/filters/FilterShorterTitles.java:10]: Add Javadoc for this class, explaining what it does, i.e. filters out titles that are shorter than the supplied statement.

File: metadata-transformer/src/main/java/no/unit/nva/metadata/filters/FilterShorterTitles.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        if (ExtractorUtil.isNotLiteral(statement.getObject())) {
+            return false;
+        }
+
+        return model.stream().filter(item -> statement.getPredicate().equals(DcTerms.TITLE.getIri()))
+                .map(Statement::getObject)
+                .filter(object -> object instanceof Literal)
+                .map(Value::stringValue)
+                .map(String::length)
+                .anyMatch(length -> length > statement.getObject().stringValue().length());
```
Comment (axthosarouris) [inline metadata-transformer/src/main/java/no/unit/nva/metadata/filters/FilterShorterTitles.java:21]: refactor to method modelContainsTitleThatIsLargerThanTheSuppliedTitleStatement

File: metadata-transformer/src/test/java/no/unit/nva/metadata/service/MetadataServiceTest.java [hunk 3]
```diff
@@ -152,7 +169,7 @@ void getCreatePublicationReturnsMostCompletePublicationDateWhenMultipleCandidate
     }
 
     @ParameterizedTest(name = "Bad date {0} is ignored in preference for shorter, valid date")
```
Comment (axthosarouris) [inline metadata-transformer/src/test/java/no/unit/nva/metadata/service/MetadataServiceTest.java:171]: The name here is wrong now. The empty string is shorter of everything else

File: metadata-transformer/src/test/java/no/unit/nva/metadata/service/testdata/TypeInformationArgumentsProvider.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.ArgumentsProvider;
+
+import java.util.stream.Stream;
+
+public class TypeInformationArgumentsProvider implements ArgumentsProvider {
+    @Override
+    public Stream<? extends Arguments> provideArguments(ExtensionContext context) throws Exception {
+        return Stream.of(Arguments.of(Citation.ISBN.getMetaTagName(), "9783110646610", Book.class, BookMonograph.class),
```
Comment (axthosarouris) [inline metadata-transformer/src/test/java/no/unit/nva/metadata/service/testdata/TypeInformationArgumentsProvider.java:17]: the ISBN and the ISSN value should be static fields  with names VALID_BOOK_ISBN (although ISBN is only for books) and the ISSN should be VALID_ISSN_FOR_JOURNAL_ARTICLE.

Comment (axthosarouris) [reply metadata-transformer/src/test/java/no/unit/nva/metadata/service/testdata/TypeInformationArgumentsProvider.java:17]: In general there is an implicit content on these numbers that should be made explicit.

File: metadata-transformer/src/test/java/no/unit/nva/metadata/service/MetadataServiceTest.java [hunk 5]
```diff
@@ -308,6 +329,99 @@ void getCreatePublicationRequestReturnsLongestTitleWhenInputContainsMultipleTitl
         assertThat(actual, equalTo(expected));
     }
 
+    @ParameterizedTest
+    @ArgumentsSource(TypeInformationArgumentsProvider.class)
+    void getCreatePublicationRequestReturnsTypeWhenInputIndicatesType(String metaTagName,
+                                                                      String metaTagContent,
```
Comment (axthosarouris) [inline metadata-transformer/src/test/java/no/unit/nva/metadata/service/MetadataServiceTest.java:335]: metaTagContent -> internationalStandardNumberImplyingContentType

File: metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/ExtractionPair.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    public boolean isAbstract() {
+        IRI predicate = statement.getPredicate();
+        return DcTerms.ABSTRACT.getIri().equals(predicate)
+                || DcTerms.DESCRIPTION.getIri().equals(predicate) && hasLiteralObject() && noAbstractProperty;
+    }
+
+    public boolean isDescription() {
+        return DcTerms.DESCRIPTION.getIri().equals(statement.getPredicate()) && hasLiteralObject()
```
Comment (axthosarouris) [inline metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/ExtractionPair.java:56]: new line in &&

**Comment by axthosarouris** [comment]
I feel that the Filters and maybe the extractors should be tested more thoroughly because there can be corner cases where they fail. The "FilterShorterTitles" especially should be tested. It is not necessary that the tests are in "FilterShorterTitlesTest". They could be in `MetadataServiceTest`. However, it would give more insight if the tests were written in a `FilterShorterTitlesTest` class.

File: metadata-transformer/src/main/java/no/unit/nva/metadata/MetadataConverter.java [hunk 7]
```diff
@@ -1,67 +1,82 @@
...
-    private URI toLexvoUri(String language) {
-        String iso3LanguageCode = ISO3_LANGUAGE_CODE_UNDEFINED;
-        if (!StringUtils.isEmpty(language)) {
-            try {
-                iso3LanguageCode = new Locale(language).getISO3Language();
-            } catch (MissingResourceException e) {
-                logger.warn("Could not map two-letter BCP-47 language code to three-letter ISO639-3 language code.", e);
-            }
-        }
+    private MetadataExtractor configureExtractor() {
```
Comment (mattiahj) [inline metadata-transformer/src/main/java/no/unit/nva/metadata/MetadataConverter.java:66]: Maybe call this getConfiguredMetadataExtractor, getMetadataExtractor or getDefaultMetadataExtractor to be consistent?

File: metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/AbstractExtractor.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+import no.unit.nva.model.EntityDescription;
+import nva.commons.core.JacocoGenerated;
+
+import java.util.function.Function;
+
+public final class AbstractExtractor {
+
+    public static final Function<ExtractionPair, EntityDescription> apply = AbstractExtractor::extract;
+
+    @JacocoGenerated
```
Comment (mattiahj) [inline metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/AbstractExtractor.java:12]: code coverage "cheating" or a good reason for this?

File: metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/ContributorExtractor.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+import java.util.function.Function;
+
+import static java.util.Objects.isNull;
+
+public final class ContributorExtractor {
+    private static final Logger logger = LoggerFactory.getLogger(ContributorExtractor.class);
+
+    public static final Function<ExtractionPair, EntityDescription> apply = extractOrConsumeError();
+
+    @JacocoGenerated
```
Comment (mattiahj) [inline metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/ContributorExtractor.java:24]: I don't mind these if they're necessary, are they?

File: metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/ContributorExtractor.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    private ContributorExtractor() {
+
+    }
+
+    private static Function<ExtractionPair, EntityDescription> extractOrConsumeError() {
```
Comment (mattiahj) [inline metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/ContributorExtractor.java:29]: While I don't object to this code, it took longer to read than a method with one or two if statement would have.

File: metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/AbstractExtractor.java (hunk 8 above)
Comment (brinxmat) [reply metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/AbstractExtractor.java:12]: Without the constructor, Jacoco claims the class is uncovered because the default constructor is uncovered

File: metadata-transformer/src/main/java/no/unit/nva/metadata/MetadataConverter.java (hunk 7 above)
Comment (brinxmat) [reply metadata-transformer/src/main/java/no/unit/nva/metadata/MetadataConverter.java:66]: I think there is a consistent naming problem, I will rethink this. In general, I like to retain "get/set" for getting/setting fields, though above, I have "getCreatePublicationRequest" because I guess I thought "createCreatePublicationRequest" sounded ridiculous.

File: metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/AbstractExtractor.java (hunk 8 above)
Comment (axthosarouris) [reply metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/AbstractExtractor.java:12]: I do not think that @JacocoGenerated is necessary. Jacoco ignores these constructors

File: metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/DateExtractor.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    private static EntityDescription extract(ExtractionPair extractionPair) {
+        if (extractionPair.isDate()) {
+            addDate(extractionPair);
+        }
+        return extractionPair.getEntityDescription();
+    }
+
+    @SuppressWarnings("PMD.CloseResource")
```
Comment (axthosarouris) [inline metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/DateExtractor.java:33]: Is it possible to use try-with-resources to avoid the warning?

File: metadata-transformer/src/main/java/no/unit/nva/metadata/filters/FilterShorterTitles.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+        return getTitleLength(documentModel).map(length -> selectLongestTitle(statement, length))
+                .orElse(false);
+    }
+
+    private static Optional<String> getTitleLength(Model documentModel) {
+        return documentModel.stream()
+                .filter(FilterShorterTitles::isTitle)
+                .map(Statement::getObject)
+                .map(Value::stringValue)
+                .findAny();
```
Comment (axthosarouris) [inline metadata-transformer/src/main/java/no/unit/nva/metadata/filters/FilterShorterTitles.java:25]: Do we expect a single title here? Should we use SingletonCollector?

File: metadata-transformer/src/main/java/no/unit/nva/metadata/filters/FilterShorterTitles.java [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+import org.eclipse.rdf4j.model.Value;
+
+import java.util.Optional;
+
+/**
+ * Filters triples with property dcterms:title from the document model by length comparison, keeping the longest title.
+ */
+public final class FilterShorterTitles {
+    public static boolean apply(Model documentModel, Statement statement) {
+        return getTitleLength(documentModel).map(length -> selectLongestTitle(statement, length))
```
Comment (axthosarouris) [inline metadata-transformer/src/main/java/no/unit/nva/metadata/filters/FilterShorterTitles.java:16]: `length` should be `title`

**Review by axthosarouris — CHANGES_REQUESTED** [review]
Only the typo where "length" should be "title"

File: metadata-transformer/src/main/java/no/unit/nva/metadata/MetadataConverter.java (hunk 7 above)
Comment (mattiahj) [reply metadata-transformer/src/main/java/no/unit/nva/metadata/MetadataConverter.java:66]: I would also prefer something without get/set, but I care more that we're consistent with whatever we land on.

Comment (mattiahj) [reply metadata-transformer/src/main/java/no/unit/nva/metadata/MetadataConverter.java:66]: What about calling them NewPublicationRequest and then have createNewPublicationRequest()

the ambiguity about whether you're creating new requests, publications or publicationrequests remains, but at least it flows of the tongue ;)

File: metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/AbstractExtractor.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    private static EntityDescription addAbstract(EntityDescription entityDescription, Statement statement) {
+        Value object = statement.getObject();
+        if (!ExtractorUtil.isNotLiteral(object)) {
+            entityDescription.setAbstract(object.stringValue());
+        }
+        return entityDescription;
+    }
+
+    private static boolean isAbstract(IRI candidate, boolean noAbstract) {
+        return DcTerms.ABSTRACT.getIri().equals(candidate)
```
[the remark this answers is not in the document]
Comment (mattiahj) [reply metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/AbstractExtractor.java]: if parenthesis help clarify for people I don't mind having them

File: metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/DescriptionExtractor.java [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    private DescriptionExtractor() {
+
+    }
+
+    @SuppressWarnings("PMD.CloseResource")
+    private static EntityDescription extract(ExtractionPair extractionPair) {
+        Statement statement = extractionPair.getStatement();
+        EntityDescription entityDescription = extractionPair.getEntityDescription();
+        if (isDescription(statement.getPredicate(), extractionPair.isNoAbstract())) {
+            addDescription(entityDescription, statement);
```
[the remark this answers is not in the document]
Comment (mattiahj) [reply metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/DescriptionExtractor.java]: +1

File: metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/DocumentTypeExtractor.java [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+
+    }
+
+    private static Function<ExtractionPair, EntityDescription> extractOrConsumeError() {
+        return (extractionPair) -> {
+            try {
+                return extract(extractionPair);
+            } catch (InvalidIssnException | InvalidIsbnException e) {
+                logger.warn("Could not extract type metadata from statement " + extractionPair.getStatement());
+                return extractionPair.getEntityDescription();
```
[the remark this answers is not in the document]
Comment (mattiahj) [reply metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/DocumentTypeExtractor.java]: I'm not really against it, but unless this returnDefaultValue would be used multiple places I don't think it is necessary. If this refactoring is done the name should be something descriptive as well, so we don't lose the context of the error, what is being logged, and what is being returned behind the method name (ie: I don't want to have to go look for a single use method to find out what the parent method actually does).

File: metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/DateExtractor.java [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+        String[] dateParts = date.split(DATE_SEPARATOR);
+        int length = dateParts.length;
+
+        PublicationDate.Builder publicationDateBuilder = new PublicationDate.Builder()
+                .withYear(dateParts[0]);
+
+        if (length > YEAR_ONLY) {
+            publicationDateBuilder.withMonth(dateParts[1]);
+        }
+        if (length == FULL_DATE) {
```
Comment (mattiahj) [inline metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/DateExtractor.java:52]: I find the use of this constant more confusing than helpful. I have to scroll up and confirm that it is a number, then scroll back down and remember the number to see it in the context of the code. Changing the name of the variable length to dateLength or dateFormatLength, or maybe introducing some kind of data format type would probably help a little too, but we might want to consider whether always making constants out of things is beneficial.

All of this is subjective, but in this specific context I personally find the dataParts indexes more "magic" than having the values in a simple int comparison there "on paper" as it where.

Comment (mattiahj) [reply metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/DateExtractor.java:52]: Actually, if we're going to stick with constants named after the format of the date, don't call the int anything to do with length at all. At that point it should be dateFormat.

File: metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/DocumentTypeExtractor.java [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+            throws InvalidIsbnException {
+        PublicationInstance<? extends Pages> instance = reference.getPublicationInstance();
+        PublicationContext context = reference.getPublicationContext();
+        if (hasExistingInstanceAndContext(instance, context)) {
+            List<String> existingIsbns = ((Book) context).getIsbnList();
+            List<String> isbnList = nonNull(existingIsbns) ? new ArrayList<>(existingIsbns) : new ArrayList<>();
+            addNonPreexistingIsbn(object, (Book) context, isbnList);
+        } else {
+            BookMonograph instanceType = new BookMonograph.Builder().build();
+            Book contextType = new Book.Builder().withIsbnList(List.of(object.stringValue())).build();
```
Comment (mattiahj) [inline metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/DocumentTypeExtractor.java]: De-rdf-izize this...

File: metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/TagExtractor.java [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Function;
+
+import static java.util.Objects.nonNull;
+
+public final class TagExtractor {
+    public static final Set<IRI> TAG_IRIS = Set.of(DcTerms.COVERAGE.getIri(), DcTerms.TEMPORAL.getIri(),
```
Comment (mattiahj) [inline metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/TagExtractor.java]: I would not mind some newlines to organize this set.

File: metadata-transformer/src/main/java/no/unit/nva/metadata/type/Bibo.java [hunk 20]
```diff
@@ -1,10 +1,12 @@
...
 
 import org.eclipse.rdf4j.model.IRI;
-import org.eclipse.rdf4j.model.ValueFactory;
+import org.eclipse.rdf4j.model.impl.SimpleValueFactory;
 
 public enum Bibo implements OntologyProperty {
-    DOI("doi");
+    DOI("doi"),
+    ISBN("isbn"),
+    ISSN("eissn");
```
Comment (mattiahj) [inline metadata-transformer/src/main/java/no/unit/nva/metadata/type/Bibo.java:9]: in general eissn != issn, should this be EISSN("eissn")?

File: metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/ExtractionPair.java [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+    private static final Set<IRI> CONTRIBUTOR_PROPERTIES = Set.of(DcTerms.CREATOR.getIri(),
+            DcTerms.CONTRIBUTOR.getIri());
+    public static final Set<IRI> DATE_PREDICATES = Set.of(DcTerms.DATE.getIri(), DcTerms.DATE_ACCEPTED.getIri(),
+            DcTerms.DATE_COPYRIGHTED.getIri(), DcTerms.DATE_SUBMITTED.getIri());
+    public static final String DATE_REGEX = "^[0-9]{4}(-[0-9]{2}(-[0-9]{2})?)?$";
+    public static final Set<IRI> TAG_IRIS = Set.of(DcTerms.COVERAGE.getIri(), DcTerms.TEMPORAL.getIri(),
+            DcTerms.SPATIAL.getIri(), DcTerms.SUBJECT.getIri());
+    private final Statement statement;
+    private final EntityDescription entityDescription;
+    private final boolean noAbstractProperty;
```
[the remark this answers is not in the document]
Comment (mattiahj) [reply metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/ExtractionPair.java]: +1

File: metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/DateExtractor.java (hunk 11 above)
Comment (mattiahj) [reply metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/DateExtractor.java:33]: The places this warning was used previously in this project this was not possible, but this should indeed be checked again.

**Review by mattiahj — DISMISSED** [review]

File: metadata-transformer/src/main/java/no/unit/nva/metadata/MetadataConverter.java [hunk 22]
```diff
@@ -1,67 +1,82 @@
...
-    public MetadataConverter(Model metadata, String jsonld) {
+    public MetadataConverter(Model metadata) {
         this.metadata = metadata;
-        this.jsonld = jsonld;
+        this.entityDescription = new EntityDescription();
+        this.emptyDescriptionHash = entityDescription.hashCode();
+    }
+
+    @SuppressWarnings("PMD.CloseResource")
+    public Optional<CreatePublicationRequest> extractPublicationRequest() {
```
Comment (axthosarouris) [inline metadata-transformer/src/main/java/no/unit/nva/metadata/MetadataConverter.java:36]: generatePublicationRequest

File: metadata-transformer/src/main/java/no/unit/nva/metadata/MetadataConverter.java [hunk 23]
```diff
@@ -1,67 +1,82 @@
...
 
-    private Optional<URI> getLanguageFromMetadata(Model metadata) {
-        return metadata
-            .stream()
-            .filter(this::isDcLanguage)
-            .map(Statement::getObject)
-            .map(Value::stringValue)
-            .map(this::toLexvoUri)
-            .findFirst();
+    private CreatePublicationRequest getCreatePublicationRequest() {
```
Comment (axthosarouris) [inline metadata-transformer/src/main/java/no/unit/nva/metadata/MetadataConverter.java:55]: newCreatePublicationRequest()?

File: metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/AbstractExtractor.java [hunk 24]
```diff
@@ -0,0 +1,27 @@
+package no.unit.nva.metadata.extractors;
+
+import no.unit.nva.model.EntityDescription;
+
+import java.util.function.Function;
+
+public final class AbstractExtractor {
```
Comment (axthosarouris) [inline metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/AbstractExtractor.java:7]: Can you add a small comment that this class extracts an Abstract and is not an abstract implementation of an extractor?

File: metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/DateExtractor.java (hunk 11 above)
Comment (axthosarouris) [reply metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/DateExtractor.java:31]: This suppression is not necessary here according to my IDE and PMD

File: metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/ExtractionPair.java [hunk 25]
```diff
@@ -0,0 +1,97 @@
...
+
+    public boolean hasLiteralObject() {
+        return statement.getObject() instanceof Literal;
+    }
+
+    public boolean isAbstract() {
+        IRI predicate = statement.getPredicate();
+        return hasLiteralObject() && (DcTerms.ABSTRACT.getIri().equals(predicate)
+                || DcTerms.DESCRIPTION.getIri().equals(predicate) && !abstractPropertyExists);
+    }
```
Comment (axthosarouris) [inline metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/ExtractionPair.java:53]: suggestion:

```
    public boolean isAbstract() {
        return hasLiteralObject() && (docHasAbstract() || docHasDescriptionServingAsAbstract());
    }

    private boolean docHasAbstract() {
        return DcTerms.ABSTRACT.getIri().equals(statement.getPredicate());
    }

    private boolean docHasDescriptionServingAsAbstract() {
        return DcTerms.DESCRIPTION.getIri().equals(statement.getPredicate()) && !abstractPropertyExists;
    }
```

File: metadata-transformer/src/main/java/no/unit/nva/metadata/filters/FilterShorterTitles.java [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+import org.eclipse.rdf4j.model.Value;
+
+import java.util.Optional;
+
+/**
+ * Filters triples with property dcterms:title from the document model by length comparison, keeping the longest title.
+ */
+public final class FilterShorterTitles {
+    public static boolean apply(Model documentModel, Statement statement) {
+        return findOtherTitle(documentModel)
```
Comment (axthosarouris) [inline metadata-transformer/src/main/java/no/unit/nva/metadata/filters/FilterShorterTitles.java:16]: I think this code selects randomly one of many titles and compares it with the value of the statement. If the randomly chosen title is longer than the value of the current statement the value of the statement is filtered out. 
Pay attention that there is no check whether the statement is a title. This means that an author could be filtered out by a long title.

Suggestion:

```
public final class FilterShorterTitles {

    public static boolean apply(Model documentModel, Statement statement) {
        if (isTitle(statement)) {
            String currentTitle = statement.getObject().stringValue();
            int maxTitleLength = findMaxTitleLength(documentModel);
            return currentTitle.length() < maxTitleLength;
        }
        return false;
    }

    private static int findMaxTitleLength(Model documentModel) {
        return documentModel.stream()
                   .filter(FilterShorterTitles::isTitle)
                   .map(Statement::getObject)
                   .map(Value::stringValue)
                   .map(String::length)
                   .reduce(Math::max)
                   .orElse(0);
    }

    private static boolean isTitle(Statement statement) {
        return DcTerms.TITLE.getIri().equals(statement.getPredicate())
               && statement.getObject() instanceof Literal;
    }
}
```

Comment (axthosarouris) [reply metadata-transformer/src/main/java/no/unit/nva/metadata/filters/FilterShorterTitles.java:16]: The respective test is also wrong.
suggestion:
```
 @ParameterizedTest
    @CsvSource({"long, longer, longest, longest", "longest, long, longer, longest"})
    void getCreatePublicationRequestReturnsLongestTitleWhenInputContainsMultipleTitles(String first,
                                                                                       String second,
                                                                                       String third,
                                                                                       String expected)
        throws IOException {
        String shortAuthorName = createAuthorShorterThanShortestTitle(first, second, third);
        List<MetaTagPair> metaTags = List.of(new MetaTagPair(DC_TITLE, first), new MetaTagPair(CITATION_TITLE, second),

                                             new MetaTagPair(CITATION_AUTHOR, shortAuthorName));
        CreatePublicationRequest createPublicationRequest = getCreatePublicationRequest(third, metaTags);
        String actual = createPublicationRequest.getEntityDescription().getMainTitle();
        List<String> authorNames = createPublicationRequest.getEntityDescription()
                                       .getContributors()
                                       .stream()
                                       .map(Contributor::getIdentity)
                                       .map(Identity::getName)
                                       .collect(Collectors.toList());
        assertThat(authorNames, hasItem(shortAuthorName));
        assertThat(actual, equalTo(expected));
    }

    private String createAuthorShorterThanShortestTitle(String... titles) {
        return Arrays.stream(titles).reduce(this::shortestString)
                   .map(this::leftHalf)
                   .orElseThrow();
    }

    private String leftHalf(String title) {
        return title.substring(0, title.length() / 2);
    }

    private String shortestString(String left, String right) {
        return (left.length() < right.length()) ? left : right;
    }
```

File: metadata-transformer/src/main/java/no/unit/nva/metadata/service/MetadataService.java [hunk 27]
```diff
@@ -89,9 +72,8 @@ public MetadataService(HttpClient httpClient) {
     public Optional<CreatePublicationRequest> getCreatePublicationRequest(URI uri) {
```
Comment (axthosarouris) [inline metadata-transformer/src/main/java/no/unit/nva/metadata/service/MetadataService.java:72]: generatePublicationRequest

File: metadata-transformer/src/test/java/no/unit/nva/metadata/service/MetadataServiceTest.java [hunk 28]
```diff
@@ -302,12 +332,105 @@ void getCreatePublicationRequestReturnsLongestTitleWhenInputContainsMultipleTitl
                                                                                        String third,
                                                                                        String expected)
             throws IOException {
-        List<MetaTagPair> metaTags = List.of(new MetaTagPair(DC_TITLE, first), new MetaTagPair(CITATION_TITLE, second));
+        List<MetaTagPair> metaTags = List.of(new MetaTagPair(DC_TITLE, first), new MetaTagPair(CITATION_TITLE, second),
+                new MetaTagPair(CITATION_AUTHOR, ENSURE_ONLY_FILTERING_TITLES));
```
Comment (axthosarouris) [inline metadata-transformer/src/test/java/no/unit/nva/metadata/service/MetadataServiceTest.java:336]: See my comment above for the test

**Review by axthosarouris — CHANGES_REQUESTED** [review]
Since we are mutating the document model, it could be worth testing the whole model or the whole request in each test. I am not sure about it though.

Requested changes are only the changes regarding the filter.

File: metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/DateExtractor.java (hunk 11 above)
Comment (brinxmat) [reply metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/DateExtractor.java:31]: We use a different version to Codacy

**Review by axthosarouris — APPROVED** [review]
